src/fp.rs

```diff
@@ -9,12 +9,58 @@ use subtle::{Choice, ConditionallySelectable, ConstantTimeEq, CtOption};
 
 use crate::util::{adc, mac, sbb};
 
+mod magnitudes;
+
+pub use magnitudes::*;
+
 // The internal representation of this type is six 64-bit unsigned
 // integers in little-endian order. `Fp` values are always in
 // Montgomery form; i.e., Scalar(a) = aR mod p, with R = 2^384.
 #[derive(Copy, Clone)]
 pub struct Fp([u64; 6]);
 
+#[derive(Debug)]
+pub struct FpWide([u64; 12]);
+
+impl FpWide {
+    pub const fn add(&self, rhs: &FpWide) -> FpWide {
+        let (d0, carry) = adc(self.0[0], rhs.0[0], 0);
+        let (d1, carry) = adc(self.0[1], rhs.0[1], carry);
+        let (d2, carry) = adc(self.0[2], rhs.0[2], carry);
+        let (d3, carry) = adc(self.0[3], rhs.0[3], carry);
+        let (d4, carry) = adc(self.0[4], rhs.0[4], carry);
+        let (d5, carry) = adc(self.0[5], rhs.0[5], carry);
+        let (d6, carry) = adc(self.0[6], rhs.0[6], carry);
+        let (d7, carry) = adc(self.0[7], rhs.0[7], carry);
+        let (d8, carry) = adc(self.0[8], rhs.0[8], carry);
+        let (d9, carry) = adc(self.0[9], rhs.0[9], carry);
+        let (d10, carry) = adc(self.0[10], rhs.0[10], carry);
+        let (d11, _) = adc(self.0[11], rhs.0[11], carry);
+
+        FpWide([d0, d1, d2, d3, d4, d5, d6, d7, d8, d9, d10, d11])
+    }
+
+    /// Negates an element of magnitude 1 by subtracting it from a multiple of
+    /// the prime that is larger than this element could be (p^2), increasing
+    /// the magnitude by 1.
+    pub const fn negate(&self) -> FpWide {
```

Comment:
This should be called `negate_u1`. (I would use a generic, but it's not stable in Rust yet for `const` functions...)